README
IPTables-Log version 0.0004
===========================

IPTables::Log provides methods for parsing and working with
iptables/netfilter syslog entries.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

	Carp
	NetAddr::IP
	Data::GUID

COPYRIGHT AND LICENCE

Copyright (C) 2009 by Andy Smith <andy.smith@netprojects.org.uk>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension IPTables::Log.

0.0004  Mon Dec 21 01:27:00 2009
	- Added ICMP type support.
	- Added 'parsed' accessor to IPTables::Log::Set::Record to get parse status.
	- Removed erroneous character in regex in Record.pm (make test now works).

0.0003  Tue Dec 15 01:31:00 2009
	- Fixed further dependency on NetAddr::IP.
	- Added 'no_header' option to IPTables::Log::Set::Record.

0.0002  Tue Oct 20 15:05:00 2009
	- Fixed dependencies in Makefile.PL (Closes: 50585).
	- Added get_by() to return a set indexes on a field.
	- Modified IPTables::Log::Set::Record->_process_(value|present)
		to *always* return a value.
	- Written some proper tests.
	- Other various bugfixes.

0.0001  Wed Oct 14 18:00:44 2009
	- original version; created by h2xs 1.23 with options
		-X -n IPTables::Log



lib/IPTables/Log.pm
#!/usr/bin/perl

#=======================================================================
# Log.pm / IPTables::Log
# $Id: Log.pm 17 2009-12-15 01:36:48Z andys $
# $HeadURL: https://daedalus.nocarrier.org.uk/svn/IPTables-Log/trunk/IPTables-Log/lib/IPTables/Log.pm $
# (c)2009 Andy Smith <andy.smith@netprojects.org.uk>
#-----------------------------------------------------------------------
#:Description
# This is the main IPTables::Log class.
#-----------------------------------------------------------------------
#:Synopsis
#
# use IPTables::Log;
# my $l = IPTables::Log->new;
# my $s = $l->create_set;
# my $r = $s->create_record({text => '...IN=eth0 OUT=eth1 MAC=00:...'});
# $r->parse;
# $s->add($r);
#=======================================================================

# The pod (Perl Documentation) for this module is provided inline. For a
# better-formatted version, please run:-
# $ perldoc Log.pm

=head1 NAME

IPTables::Log - Parse iptables/netfilter syslog messages.

=head1 SYNOPSIS

  use IPTables::Log;
  my $l = IPTables::Log->new;
  my $s = $l->create_set;
  my $r = $s->create_record({text => '...IN=eth0 OUT=eth1 MAC=00:...'});
  $r->parse;
  $s->add($r);
  
=head1 DEPENDENCIES

=over 4

=item * Carp - for error generation

=item * Class::Accessor - for accessor methods

=item * Data::GUID - for GUID generation

=item * NetAddr::IP - for the C<src> and C<dst> methods

=back

=cut

# Set our package name
package IPTables::Log;

# Set minimum version of Perl
use 5.010000;
# Use strict and warnings
use strict;
use warnings;

# Use Carp for errors
use Carp;
# Use IPTables::Log::Set
use IPTables::Log::Set;

# Inherit from Class::Accessor to simplify accessor methods.
use base qw(Class::Accessor);
__PACKAGE__->follow_best_practice;
__PACKAGE__->mk_accessors( qw(raw debug) );

# Set version information
our $VERSION;
$VERSION = "0.0004";

# Hashes of colour
my $clr = "[0m";
my $bold = "[1m";
my $fclr = {'red' => '[31;1m',
			'green' => '[32;1m',
			'yellow' => '[33;1m',
			'blue' => '[34;1m',
			'purple' => '[35;1m',
			'cyan' => '[36;1m'};

my $bclr = {'red' => '[41;1m',
			'green' => '[42;1m',
			'yellow' => '[43;1m',
			'blue' => '[44;1m',
			'purple' => '[45;1m',
			'cyan' => '[46;1m'};

# Generates a debug message if $self->debug == 1
sub debug
{
	my ($self, $msg) = @_;

	if($self->get_debug)
	{
		print $bclr->{blue}.$fclr->{yellow}."D".$clr." ".$fclr->{green}.__PACKAGE__.$clr." ".$fclr->{purple}.$VERSION.$clr." | ".$msg."\n";
	}
}

# As above, but doesn't append a newline
sub debug_nolf
{
	my ($self, $msg) = @_;

	if($self->get_debug)
	{
		print $bclr->{blue}.$fclr->{yellow}."D".$clr." ".$msg;
	}
}

# As per $self->debug, but prints additional information in a chosen colour
sub debug_value
{
	my ($self, $text, $colour, $value) = @_;

	$self->debug($text." ".$self->fcolour($colour, $value));
}

# Prints an error to STDERR
sub error
{
	my ($self, $msg) = @_;

	print STDERR $fclr->{red}."E".$clr." ".$msg."\n";
}

# Prints and error to STDERR, then 'croak's
sub fatal
{
	my ($self, $msg) = @_;

	croak $bclr->{red}.$bold."!".$clr." ".$msg."\n";
}

# Wrap given message in ANSI colour codes
sub fcolour
{
	my ($self, $colour, $text) = @_;

	return $fclr->{$colour}.$text.$clr;
}

=head1 CONSTRUCTORS

=head2 Log->new

Creates a new C<IPTables::Log> object.

=head1 METHODS

=head2 $log->create_set(I<no_header => 0|1>)

Creates a new C<IPTables::Log::Set> object.

Setting I<no_header> to B<1> makes L<IPTables::Log::Set::Record> assume that the timestamp and hostname at the beginning of the message is missing (for example, if it's already been processed by another utility).

See L<IPTables::Log::Set> and L<IPTables::Log::Set::Record> for further details.

=cut

sub create_set
{
	my ($self, $args) = @_;

	$args->{'log'} = $self;

	my $set = IPTables::Log::Set->create($args);

	return $set;
}

=head1 CAVEATS

It parses log entries. It doesn't do much else, yet.

=head1 BUGS

None that I'm aware of ;-)

=head1 AUTHOR

This module was written by B<Andy Smith> <andy.smith@netprojects.org.uk>.

=head1 COPYRIGHT

$Id: Log.pm 17 2009-12-15 01:36:48Z andys $

(c)2009 Andy Smith (L<http://andys.org.uk/>)

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1


lib/IPTables/Log/Set.pm
#!/usr/bin/perl

#=======================================================================
# Set.pm / IPTables::Log::Set
# $Id: Set.pm 17 2009-12-15 01:36:48Z andys $
# $HeadURL: https://daedalus.nocarrier.org.uk/svn/IPTables-Log/trunk/IPTables-Log/lib/IPTables/Log/Set.pm $
# (c)2009 Andy Smith <andy.smith@netprojects.org.uk>
#-----------------------------------------------------------------------
#:Description
# This class holds a set of IPTables::Log::Set::Record objects
#-----------------------------------------------------------------------
#:Synopsis
# NOTE: This class isn't designed to be created directly.
#
# use IPTables::Log;
# my $l = IPTables::Log->new;
# my $s = $l->create_set;
# my $r = $s->create_record({text => '...IN=eth0 OUT=eth1 MAC=00:...'});
# $r->parse;
# $s->add($r);
#=======================================================================

# The pod (Perl Documentation) for this module is provided inline. For a
# better-formatted version, please run:-
# $ perldoc Set.pm

=head1 NAME

IPTables::Log::Set - Holds a set of IPTables::Log::Set::Record objects.

=head1 SYNOPSIS

Note that this class isn't designed to be created directly. You can create these objects via a C<IPTables::Log> object.

  use IPTables::Log;
  my $l = IPTables::Log->new;
  my $s = $l->create_set;

=head1 DEPENDENCIES

=over 4

=item * Class::Accessor - for accessor methods

=item * Data::GUID - for GUID generation

=item * NetAddr::IP - for the C<src> and C<dst> methods (required by L<IPTables::Log::Set::Record>)

=back

=cut

# Set our package name
package IPTables::Log::Set;

# Minimum version
use 5.010000;
# Use strict and warnings
use strict;
use warnings;

# Use Carp for erroring
use Carp;
# Use Data::GUID for generating GUIDs
use Data::GUID;
# Use IPTables::Log::Set::Record for individual log entries
use IPTables::Log::Set::Record;
# Use Data::Dumper
use Data::Dumper;

# Inherit from Class::Accessor to simplify accessor method generation
use base qw(Class::Accessor);
# Follow best practice
__PACKAGE__->follow_best_practice;
# Create log and guid as read-only accessor methods
__PACKAGE__->mk_ro_accessors( qw(log guid) );

# Set version information
our $VERSION = '0.0004';

=head1 CONSTRUCTORS

=head2 Set->create

Creates a new C<IPTables::Log::Set> object. This isn't the recommended way to do this, however. The proper way is to create an object via a L<IPTables::Log> object with C<create_set>.

=cut

sub create
{
	my ($class, $args) = @_;

	my $self = __PACKAGE__->new($args);
	$self->{records} = {};

	# Generate a GUID for the set
	my $g = Data::GUID->new;
	$self->{guid} = $g->as_string;
	$self->{no_header} = $args->{'no_header'};

	return $self;
}

=head1 METHODS

=head2 $set->create_record(I<{text => '...IN=eth0 OUT=eth1 MAC=00:...'}>))

Creates a new L<IPTables::Log::Set::Record> object. This is the B<recommended> way to create C<IPTables::Log::Set::Record> objects, as it ensures various settings are inherited from the C<Log> class.

The text of the log entry can be passed here, or it can be passed with the C<set_text> accessor method to the C<IPTables::Log::Set::Record> object itself.

=cut

sub create_record
{
	my ($self, $args) = @_;

	#$args->{log} = $self->get_log;

	my $record = IPTables::Log::Set::Record->create($args);

	return $record;
}

=head2 $set->load_file($filename)

Loads in logs from I<$filename>, discarding any which don't appear to be iptables/netfilter logs. A L<IPTables::Log::Set::Record> object is then created for each entry, and the content is then parsed. Finally, each entry is then added to the set created with C<create_record>.

=cut

sub load_file
{
	my ($self, $filename) = @_;

	# Check we've been passed a filename
	if(!$filename)
	{
		croak "No filename given to load_file().";
		#$self->get_log->fatal("No filename given!");
	}

	# Check that the file exists, and barf if not.
	if(!-f $filename)
	{
		croak $filename." does not exist.";
		#$self->get_log->fatal("Cannot find ".$self->get_log->fcolour('yellow', $filename));
	}

	#$self->get_log->debug("Opening ".$self->get_log->fcolour('yellow', $filename)."...");
	# Open the logfile
	open(LOGFILE, $filename) || $self->get_log->fatal("Cannot open ".$self->get_log->fcolour('yellow', $filename));
	my @logs = <LOGFILE>;
	#$self->get_log->debug("Finished reading in logs.");

	# It's a fair bet that if we don't have an IN= and an OUT= and it doesn't have a source of 'kernel', then it's not an iptables log.
	# We'll discard those before even attempting to parse it.
	foreach my $log (@logs)
	{
		if($log =~ /kernel.+IN=.+OUT=/)
		{
			chomp($log);
			#$self->get_log->debug_nolf("Parsing iptables log entry... ");
			my $record = $self->create_record({'text' => $log, 'no_header' => $self->{no_header}});
			$record->parse;
			#$self->get_log->debug("done.");
			$self->add($record);
			$self->get_log->debug("Added record with GUID ".$self->get_log->fcolour('yellow', $record->get_guid). " to set.");
			#return 1;
		}
		else
		{
			#$self->get_log->debug("Log entry is not an iptables log entry, so skipping...");
		}
	}
	return 1;
}

=head2 $set->add($record)

Adds a L<IPTables::Log::Set::Record> object to a set created with C<create_set>.

=cut

sub add
{
	my ($self, $record) = @_;

	if($record)
	{
		my $guid = $record->get_guid;

		$self->{records}{$guid} = $record;
	}
}

=head2 $set->get_by('field')

Returns a hash of record identifiers, indexed by I<field>. Field can be one of I<guid>, I<date>, I<time>, I<hostname>, I<prefix>, I<in>, I<out>, I<mac>, I<src>, I<dst>, I<proto>, I<spt>, I<dpt>, I<id>, I<len>, I<ttl>, I<df>, I<window>, I<syn>.

If you attempt to sort on a field that isn't present in all records in the set, get_by will only return records which have that field. For example, if you attempt to get_by('dpt'), any ICMP log messages will be silently excluded from the returned set.

=cut

sub get_by
{
	my ($self, $by) = @_;

	# Check that $by is set
	if($by)
	{
		# Create a hash to hold the index values
		my %indexes;
		$indexes{by} = $by;

		foreach my $r (keys %{$self->{records}})
		{
			# Step through each record.
			my $record = $self->{records}{$r};
			my $value = $record->get($by);

			# If $value is blank, it means not all records have this field.
			# For now, we'll refuse to add these.
			if($value)
			{
				if(!$indexes{$by}{$record->get($by)})
				{
					$indexes{$by}{$record->get($by)} = [];
				}
				push (@{$indexes{$by}{$record->get($by)}}, $record);
			}
		}

		return %indexes;
	}
}

=head1 CAVEATS

None.

=head1 BUGS

None that I'm aware of ;-)

=head1 AUTHOR

This module was written by B<Andy Smith> <andy.smith@netprojects.org.uk>.

=head1 COPYRIGHT

$Id: Set.pm 17 2009-12-15 01:36:48Z andys $

(c)2009 Andy Smith (L<http://andys.org.uk/>)

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1


lib/IPTables/Log/Set/Record.pm
#!/usr/bin/perl

#=======================================================================
# Record.pm / IPTables::Log::Set::Record
# $Id: Record.pm 18 2009-12-16 00:05:52Z andys $
# $HeadURL: https://daedalus.nocarrier.org.uk/svn/IPTables-Log/trunk/IPTables-Log/lib/IPTables/Log/Set/Record.pm $
# (c)2009 Andy Smith <andy.smith@netprojects.org.uk>
#-----------------------------------------------------------------------
#:Description
# This class holds a single IPTables/Netfilter record.
#-----------------------------------------------------------------------
#:Synopsis
# NOTE: This class isn't designed to be created directly.
#
# use IPTables::Log;
# my $l = IPTables::Log->new;
# my $s = $l->create_set;
# my $r = $s->create_record({text => '...IN=eth0 OUT=eth1 MAC=00:...'});
# $r->parse;
#=======================================================================

# The pod (Perl Documentation) for this module is provided inline. For a
# better-formatted version, please run:-
# $ perldoc Record.pm

=head1 NAME

IPTables::Log::Set::Record - Holds a single IPTables/Netfilter log entry.

=head1 SYNOPSIS

Note that this class isn't designed to be created directly. You can create these objects via a C<IPTables::Log::Set> object.

  use IPTables::Log;
  my $l = IPTables::Log->new;
  my $s = $l->create_set;
  my $r = $s->create_record({text => '...IN=eth0 OUT=eth1 MAC=00:...'});
  $r->parse;

=head1 DEPENDENCIES

=over 4

=item * Class::Accessor - for accessor methods

=item * Data::GUID - for GUID generation

=item * NetAddr::IP - for the C<src> and C<dst> methods

=back

=cut

# Set the package name
package IPTables::Log::Set::Record;

use 5.010000;
use strict;
use warnings;

# Use Data::GUID for generating GUIDs
use Data::GUID;
# Use Data::Dumper for debugging. Can be removed for releases.
use Data::Dumper;
# Use NetAddr::IP for IP addresses
use NetAddr::IP;
use NetAddr::IP::Util qw(inet_aton);

# Inherit from Class::Accessor, which saves us quite a bit of time.
use base qw(Class::Accessor);
# Follow best practice
__PACKAGE__->follow_best_practice;
# Make 'text' a read/write accessor method
__PACKAGE__->mk_accessors( qw(text parsed) );
# Make the rest read-only
__PACKAGE__->mk_ro_accessors( qw(log guid date time hostname prefix in out mac src dst proto _spt _dpt spt dpt id len ttl df window syn type code) );

# Set version information
our $VERSION = '0.0004';

=head1 CONSTRUCTORS

=head2 Record->create(I<{text => '...IN=eth0 OUT=eth1 MAC=00:...'}>)

Creates a new C<IPTables::Log::Set::Record> object. You shouldn't call this directly - see the synopsis for an example.

=cut

# Call create instead of new, and the GUID will be generated automatically
sub create
{
	my ($class, $args) = @_;

	my $self = __PACKAGE__->new($args);
	# Generate a GUID for the ID
	my $g = Data::GUID->new;
	$self->{guid} = $g->as_string;
	$self->{no_header} = $args->{'no_header'} ? $args->{'no_header'} : 0;
	$self->{parsed} = 0;

	return $self;
}

# Private function for checking the content of fields
# Not documented in pod format because this is a private function.
sub _process_value
{
	my ($self, $value, $name) = @_;

	# If $value isn't set, set it to "NONE". A blank string will break IPTables::Log::Set->get_by().
	if((!$value) || ($value eq ""))
	{
		$value = "NONE";
	}

	$self->{$name} = $value;
	return 1;
}

# As for _process_value, but if true replaces the value with a 1, otherwise replaces it with a 0
# Not documented in pod format because this is a private function.
sub _process_present
{
	my ($self, $value, $name) = @_;

	if($value)
	{
		$self->{$name} = 1;
	}
	else
	{
		$self->{$name} = 0;
	}
	return 1;
}

=head1 METHODS

=head2 $record->parse

Parses the log message text passed either to the constructor, or via C<set_text>.

=cut

# Parses the log text
sub parse
{
	my ($self, $text) = @_;

	if(!$self->get_text)
	{
		if($text)
		{
			# Set the text attribute to the original log message
			$self->set_text($text);
		}
		else
		{	
			#$self->get_log->error("No log text found?");
			return;
		}
	}
	else
	{
		$text = $self->get_text;
	}

	#$self->get_log->debug_value("Original log message is", 'yellow', $text);

	# First, we pull parts out common to all protocols
	my ($date, $time, $hostname, $prefix, $in, $out, $mac, $src, $dst, $len, $ttl, $id, $df, $proto);
	if($self->{'no_header'} eq 1)
	{
		(undef, $prefix, $in, $out, undef, $mac, $src, $dst, $len, $ttl, $id, $df, $proto)
			= $text =~ /kernel:(\s\[\d+\.\d+\])?\s(\S*)\sIN=(\S*)\sOUT=(\S*)\s(MAC=)?(\S+)?\s*SRC=(\d+\.\d+\.\d+\.\d+|\S+)\sDST=(\d+\.\d+\.\d+\.\d+|\S+)\sLEN=(\d+).+TTL=(\d+).+ID=(\d+)\s(DF)*\s*PROTO=(\S+)/;
	}
	else
	{
		($date, $time, $hostname, undef, $prefix, $in, $out, undef, $mac, $src, $dst, $len, $ttl, $id, $df, $proto)
			= $text =~ /(\w{3}\s\d{1,2})\s{1,2}(\d{2}:\d{2}:\d{2})\s(.+)\skernel:(\s\[\d+\.\d+\])?\s(\S*)\sIN=(\S*)\sOUT=(\S*)\s(MAC=)?(\S+)?\s*SRC=(\d+\.\d+\.\d+\.\d+|\S+)\sDST=(\d+\.\d+\.\d+\.\d+|\S+)\sLEN=(\d+).+TTL=(\d+).+ID=(\d+)\s(DF)*\s*PROTO=(\S+)/;
	}

	# Get the protocol first. Based on this, we know what regex we need next.
	$self->_process_value($proto, 'proto');
	if(!$proto)
	{
		#$self->get_log->error("Cannot determine the protocol for this log message!");
		#$self->get_log->error("The log text is ".$self->get_log->fcolour('yellow', $text));
		return;
	}

	# Process values
	# Date
	$self->_process_value($date, 'date');
	# Time
	$self->_process_value($time, 'time');
	# Hostname
	$self->_process_value($hostname, 'hostname');
	# IPTable logging prefix (as specified by '-j LOG --log-prefix="<prefix>"'
	$self->_process_value($prefix, 'prefix');
	# Ingress interface
	$self->_process_value($in, 'in');
	# Egress interface
	$self->_process_value($out, 'out');
	# MAC address, if applicable
	$self->_process_value($mac, 'mac');
	# Source IP
	$self->_process_value($src, '_src');
	if($self->{_src})
	{
		$self->{_src} = new_from_aton NetAddr::IP::Lite (inet_aton($self->{_src}));
		$self->{src} = $self->{_src}->addr();
	}
	# Destination IP
	$self->_process_value($dst, '_dst');
	if($self->{_dst})
	{
		$self->{_dst} = new_from_aton NetAddr::IP::Lite (inet_aton($self->{_dst}));
		$self->{dst} = $self->{_dst}->addr();
	}
	# Packet length
	$self->_process_value($len, 'len');
	# TTL
	$self->_process_value($ttl, 'ttl');
	# Packet ID
	$self->_process_value($id, 'id');
	# Don't fragment
	$self->_process_present($df, 'df');

	if(($proto eq "TCP") || ($proto eq "UDP"))
	{
		# TCP or UDP packet
		my ($spt, $dpt) = $text =~ /PROTO=$proto\sSPT=(\d+)\sDPT=(\d+)/;

		# Source port
		$self->_process_value($spt, 'spt');
		# Destination port
		$self->_process_value($dpt, 'dpt');

		if($proto eq "TCP")
		{
			# TCP specifics
			my ($window, $syn) = $text =~ /WINDOW=(\d+).*(SYN)/;

			# TCP window size
			$self->_process_value($window, 'window');
			# SYN present?
			$self->_process_present($syn, 'syn');
		}
	}
	elsif($proto eq "ICMP")
	{
		my ($type) = $text =~ /ICMP TYPE=(\d+)\sCODE=(\d+)/;

		# ICMP Type
		$self->_process_value($type, 'type');
		$self->_process_value($type, 'code');
	}

	# Return true if we've gotten this far.
	$self->set_parsed(1);
	return 1;
}

=head2 $record->set_text("...IN=eth0 OUT=eth1 MAC=00:...")

Sets the log message text. Either this must be set, or the text must have been passed to C<create>, otherwise C<parse> will error.

=head1 ACCESSOR METHODS

=head2 get(I<field>)

Returns the value of I<field>. Field can be one of I<guid>, I<date>, I<time>, I<hostname>, I<prefix>, I<in>, I<out>, I<mac>, I<src>, I<dst>, I<proto>, I<spt>, I<dpt>, I<id>, I<len>, I<ttl>, I<df>, I<window>, I<syn>.

=cut

# Get accessor that takes the variable to return as an argument
sub get
{
	my ($self, $value) = @_;

	return $self->{$value};
}

=head2 get_guid

Returns the GUID for the packet.

=head2 get_date

Returns the date portion of the log message.

=head2 get_time

Returns the time portion of the log message.

=head2 get_hostname

rETURns the hostname portion of the log message.

=head2 get_prefix

Returns the iptables/netfilter log prefix for the log message, i.e. the part specified by C<-j LOG --log-prefix='I<LOG PREFIX> '>.

=head2 get_in

Returns the ingress interface, if specified.

=head2 get_out

Returns the egress interface, if specified.

=head2 get_mac

Returns the MAC address, if specified.

=head2 get_src

Returns the source IP address.

=head2 get_dst

Returns the destination IP address.

=head2 get_proto

Returns the protocol.

=head2 get_spt - TCP and UDP packets only.

Returns the source port, if applicable.

=head2 get_dpt - TCP and UDP packets only.

Returns the destination port, if applicable

=head2 get_id

Returns the packet ID.

=head2 get_len

Returns the packet length.

=head2 get_ttl

Returns the packet's TTL (Time To Live).

=head2 get_df

Returns the packet's DF (Don't Fragment) value.

=head2 get_window - TCP and UDP packets only.

Returns the packet's window size.

=head2 get_sync

Returns 1 if the packet is a SYN, otherwise returns 0.

=head2 get_parsed

Returns 1 if the packet has been successfully parsed, otherwise returns 0.

=head1 CAVEATS

It parses log entries. It doesn't do much else, yet.

=head1 BUGS

None that I'm aware of ;-)

=head1 AUTHOR

This module was written by B<Andy Smith> <andy.smith@netprojects.org.uk>.

=head1 COPYRIGHT

$Id: Record.pm 18 2009-12-16 00:05:52Z andys $

(c)2009 Andy Smith (L<http://andys.org.uk/>)

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1
